File: Assets/Bundles/Client/Resources/TS/Controllers/Global/Camera/DefaultCameraModes/FlyCameraMode.lua
-- Compiled with unity-ts v2.1.0-75
local _UserInput = require("Shared/TS/UserInput/init")
local Keyboard = _UserInput.Keyboard
local Mouse = _UserInput.Mouse
local Bin = require("Shared/TS/Util/Bin").Bin
local Spring = require("Shared/TS/Util/Spring").Spring
local CameraTransform = require("Client/TS/Controllers/Global/Camera/CameraTransform").CameraTransform
local SPEED = 12
local ROTATION_SENSITIVITY = 0.005
local MIN_ROT_X = math.rad(1)
local MAX_ROT_X = math.rad(179)
local FlyCameraMode
do
	FlyCameraMode = setmetatable({}, {
		__tostring = function()
			return "FlyCameraMode"
		end,
	})
	FlyCameraMode.__index = FlyCameraMode
	function FlyCameraMode.new(...)
		local self = setmetatable({}, FlyCameraMode)
		return self:constructor(...) or self
	end
	function FlyCameraMode:constructor()
		self.bin = Bin.new()
		self.xRot = math.rad(90)
		self.yRot = 0
		self.keysDown = {}
		self.rightClicking = false
	end
	function FlyCameraMode:OnStart(camera)
		local transform = camera.transform
		self.positionSpring = Spring.new(transform.position, 5)
		self.xRotSpring = Spring.new(Vector3.new(math.rad(90), 0, 0), 5)
		self.yRotVelSpring = Spring.new(Vector3.new(0, 0, 0), 3)
		self.keyboard = self.bin:Add(Keyboard.new())
		self.mouse = self.bin:Add(Mouse.new())
		-- Sink keys:
		local sinkKeys = {}
		sinkKeys[37] = true
		sinkKeys[15] = true
		sinkKeys[33] = true
		sinkKeys[18] = true
		sinkKeys[63] = true
		sinkKeys[64] = true
		sinkKeys[61] = true
		sinkKeys[62] = true
		sinkKeys[31] = true
		sinkKeys[19] = true
		self.bin:Add(self.keyboard.KeyDown:ConnectWithPriority(0, function(event)
			local _key = event.Key
			if sinkKeys[_key] ~= nil then
				local _keysDown = self.keysDown
				local _key_1 = event.Key
				_keysDown[_key_1] = true
				event:SetCancelled(true)
			end
		end))
		self.bin:Add(self.keyboard.KeyUp:ConnectWithPriority(0, function(event)
			local _key = event.Key
			if sinkKeys[_key] ~= nil then
				local _keysDown = self.keysDown
				local _key_1 = event.Key
				_keysDown[_key_1] = nil
			end
		end))
	end
	function FlyCameraMode:OnStop()
		self.bin:Clean()
	end
	function FlyCameraMode:OnUpdate(dt)
		-- Input:
		local up = self.keysDown[37] ~= nil or self.keysDown[63] ~= nil
		local dn = self.keysDown[33] ~= nil or self.keysDown[64] ~= nil
		local lf = self.keysDown[15] ~= nil or self.keysDown[61] ~= nil
		local rt = self.keysDown[18] ~= nil or self.keysDown[62] ~= nil
		local q = self.keysDown[31] ~= nil
		local e = self.keysDown[19] ~= nil
		local direction = self:CalculateDirection()
		-- Build move vector:
		local moveVector = Vector3.new(0, 0, 0)
		if up ~= dn then
			local _moveVector = moveVector
			local _arg0 = if dn then 1 else -1
			local _arg0_1 = direction * _arg0
			moveVector = _moveVector + _arg0_1
		end
		if lf ~= rt then
			local perpendicular = direction:Cross(Vector3.up)
			local _moveVector = moveVector
			local _arg0 = if rt then 1 else -1
			local _arg0_1 = perpendicular * _arg0
			moveVector = _moveVector + _arg0_1
		end
		if q ~= e then
			local perpendicular = direction:Cross(Vector3.up)
			local upwards = direction:Cross(perpendicular)
			local _moveVector = moveVector
			local _arg0 = if q then 1 else -1
			local _arg0_1 = upwards * _arg0
			moveVector = _moveVector + _arg0_1
		end
		-- Set new position:
		if moveVector.sqrMagnitude > 0 then
			local _goal = self.positionSpring.goal
			local _normalized = moveVector.normalized
			local _arg0 = SPEED * dt
			self.positionSpring.goal = _goal + (_normalized * _arg0)
		end
		-- Handle camera rotation when right-clicking:
		local rightClick = self.mouse:IsRightButtonDown()
		self.rightClicking = rightClick
		if rightClick then
			local mouseDelta = self.mouse:GetDelta()
			self.xRotSpring.goal = Vector3.new(math.clamp(self.xRotSpring.goal.x + mouseDelta.y * ROTATION_SENSITIVITY, MIN_ROT_X, MAX_ROT_X), 0, 0)
			self.xRot = self.xRotSpring:update(dt).x
			self.yRotVelSpring.goal = Vector3.new(0, -mouseDelta.x * ROTATION_SENSITIVITY, 0)
		else
			self.yRotVelSpring.goal = Vector3.new(0, 0, 0)
		end
		self.yRot = (self.yRot + self.yRotVelSpring:update(dt).y) % (math.pi * 2)
	end
	function FlyCameraMode:OnPostUpdate()
	end
	function FlyCameraMode:OnLateUpdate(dt)
		local position = self.positionSpring:update(dt)
		local rotation = Quaternion.Euler(-self.xRot + math.pi / 2, -self.yRot, 0)
		return CameraTransform.new(position, rotation)
	end
	function FlyCameraMode:CalculateDirection()
		local yRot = self.yRot - math.pi / 2
		local xPos = math.cos(yRot) * math.sin(self.xRot)
		local zPos = math.sin(yRot) * math.sin(self.xRot)
		local yPos = math.cos(self.xRot)
		return Vector3.new(xPos, yPos, zPos)
	end
end
return {
	FlyCameraMode = FlyCameraMode,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Client/Controllers/Global/Camera/DefaultCameraModes/FlyCameraMode.ts
import { Keyboard, Mouse } from "Shared/UserInput";
import { Bin } from "Shared/Util/Bin";
import { Signal, SignalPriority } from "Shared/Util/Signal";
import { Spring } from "Shared/Util/Spring";
import { CameraMode } from "../CameraMode";
import { CameraTransform } from "../CameraTransform";

const SPEED = 12;
const ROTATION_SENSITIVITY = 0.005;

const MIN_ROT_X = math.rad(1);
const MAX_ROT_X = math.rad(179);

export class FlyCameraMode implements CameraMode {
	private bin = new Bin();

	private xRot = math.rad(90);
	private yRot = 0;

	private positionSpring!: Spring;
	private xRotSpring!: Spring;
	private yRotVelSpring!: Spring;

	private keyboard!: Keyboard;
	private mouse!: Mouse;
	private readonly keysDown = new Set<KeyCode>();

	private rightClicking = false;

	OnStart(camera: Camera) {
		const transform = camera.transform;
		this.positionSpring = new Spring(transform.position, 5);
		this.xRotSpring = new Spring(new Vector3(math.rad(90), 0, 0), 5);
		this.yRotVelSpring = new Spring(new Vector3(0, 0, 0), 3);

		this.keyboard = this.bin.Add(new Keyboard());
		this.mouse = this.bin.Add(new Mouse());

		// Sink keys:
		const sinkKeys = new Set<KeyCode>();
		sinkKeys.add(KeyCode.W);
		sinkKeys.add(KeyCode.A);
		sinkKeys.add(KeyCode.S);
		sinkKeys.add(KeyCode.D);
		sinkKeys.add(KeyCode.UpArrow);
		sinkKeys.add(KeyCode.DownArrow);
		sinkKeys.add(KeyCode.LeftArrow);
		sinkKeys.add(KeyCode.RightArrow);
		sinkKeys.add(KeyCode.Q);
		sinkKeys.add(KeyCode.E);

		for (const key of sinkKeys) {
			this.bin.Add(
				this.keyboard.OnKeyDown(
					key,
					(event) => {
						this.keysDown.add(event.KeyCode);
						event.SetCancelled(true);
					},
					SignalPriority.HIGHEST,
				),
			);
			this.bin.Add(
				this.keyboard.OnKeyUp(
					key,
					(event) => {
						this.keysDown.delete(event.KeyCode);
					},
					SignalPriority.HIGHEST,
				),
			);
		}
	}

	OnStop() {
		this.bin.Clean();
	}

	OnUpdate(dt: number) {
		// Input:
		const up = this.keysDown.has(KeyCode.W) || this.keysDown.has(KeyCode.UpArrow);
		const dn = this.keysDown.has(KeyCode.S) || this.keysDown.has(KeyCode.DownArrow);
		const lf = this.keysDown.has(KeyCode.A) || this.keysDown.has(KeyCode.LeftArrow);
		const rt = this.keysDown.has(KeyCode.D) || this.keysDown.has(KeyCode.RightArrow);
		const q = this.keysDown.has(KeyCode.Q);
		const e = this.keysDown.has(KeyCode.E);

		const direction = this.CalculateDirection();

		// Build move vector:
		let moveVector = new Vector3(0, 0, 0);
		if (up !== dn) {
			moveVector = moveVector.add(direction.mul(dn ? 1 : -1));
		}
		if (lf !== rt) {
			const perpendicular = direction.Cross(Vector3.up);
			moveVector = moveVector.add(perpendicular.mul(rt ? 1 : -1));
		}
		if (q !== e) {
			const perpendicular = direction.Cross(Vector3.up);
			const upwards = direction.Cross(perpendicular);
			moveVector = moveVector.add(upwards.mul(q ? 1 : -1));
		}

		// Set new position:
		if (moveVector.sqrMagnitude > 0) {
			this.positionSpring.goal = this.positionSpring.goal.add(moveVector.normalized.mul(SPEED * dt));
		}

		// Handle camera rotation when right-clicking:
		const rightClick = this.mouse.IsRightButtonDown();
		this.rightClicking = rightClick;
		if (rightClick) {
			const mouseDelta = this.mouse.GetDelta();
			this.xRotSpring.goal = new Vector3(
				math.clamp(this.xRotSpring.goal.x + mouseDelta.y * ROTATION_SENSITIVITY, MIN_ROT_X, MAX_ROT_X),
				0,
				0,
			);
			this.xRot = this.xRotSpring.update(dt).x;
			this.yRotVelSpring.goal = new Vector3(0, -mouseDelta.x * ROTATION_SENSITIVITY, 0);
		} else {
			this.yRotVelSpring.goal = new Vector3(0, 0, 0);
		}
		this.yRot = (this.yRot + this.yRotVelSpring.update(dt).y) % (math.pi * 2);
	}

	OnPostUpdate() {}

	OnLateUpdate(dt: number) {
		const position = this.positionSpring.update(dt);
		const rotation = Quaternion.Euler(-this.xRot + math.pi / 2, -this.yRot, 0);
		return new CameraTransform(position, rotation);
	}

	private CalculateDirection(): Vector3 {
		const yRot = this.yRot - math.pi / 2;
		const xPos = math.cos(yRot) * math.sin(this.xRot);
		const zPos = math.sin(yRot) * math.sin(this.xRot);
		const yPos = math.cos(this.xRot);
		return new Vector3(xPos, yPos, zPos);
	}
}
